File: py/test_data_network.py
"""Module test_data_network
   Generates and plots test data for TE estimations for multiple users
"""

import numpy as np
import test_data_utils



def __lamY(t, mu, gammas, Sx_list, influencer_ids):
    """Evaluates λ_Y(t | S_(N(Y)^t))
    
    Arguments:
        t {float} -- time boundary in hours
        mu {float} -- background noise rate
        gammas {list of float} -- list of influence rates (How much does each X influence Y?)
        Sx_list {dict of int: list of float} -- timestamps Sx of activity of users X (influencers)
        influencer_ids {list of int} --- list of ids of neighbors of Y
    Returns:
        float -- value of function λ_Y
    """
    NY_sum = 0

    for i in influencer_ids:
        Sx_t = [tx for tx in Sx_list[i] if tx < t]
        g_sum = sum([test_data_utils.g(t - ti) for ti in Sx_t if ti < t])
        NY_sum += (gammas[i] * g_sum)
    
    return mu + NY_sum


def NHppp(T, lam_upper, mu, gammas, Sx_list, influencer_ids):
    """NPPP - Non-homogeneous poisson point process
    Generates data according to NPPP 

    Arguments:
        T {int} -- time boundary
        lam_upper {float} -- maximum of function lamY
        mu {float} -- background noise rate
        gammas {dict of int: float} -- list of influence rates (How much does each X influence Y?)
        Sx_list {list of list of float} -- timestamps Sx of activity of users X (influencers)
        influencer_ids {list of int} -- ids of influencers/neighbors of Y
    
    Returns:
        list -- timestamps of activity of user Y
    """
    
    t = 0
    ts = []
    while t < T:
        u = np.random.uniform()
        t = t - 1 / lam_upper * np.log(u)
        u2 = np.random.uniform()
        
        if u2 <= (__lamY(t, mu, gammas, Sx_list, influencer_ids) / lam_upper):
            if t < T:
                ts.append(t)
    return ts



def get_test_data(T, lam_Sx, mu, gammas, influencer):
    """Creates test data for a network of users
    
    Arguments:
        T {int} -- time boundary in hours
        lam_Sx {dict of int: float} -- λ used for generating data for users X (influencers) 
        mu {float} -- background noise rate
        gammas {dict of int: float} -- list of influence rates (How much does each X influence Y?)
        influencer {dict of int: list of int} -- neighbor map for the whole network mapping Y user ids (follower) to X user ids (influencer)

    Returns:
        list of list of float -- timestamps Sx of activity of users X (influencers)
        list of list of float -- timestamps Sy of activity of users Y (followers)
    """
    influencer_ids = gammas.keys()
    follower_ids = influencer.keys()

    Sx_list = {}
    for i in influencer_ids:
        Sx_list[i] = test_data_utils.Hppp(T, lam_Sx[i])

    Sy_list = {}
    for i in follower_ids:
        lam_upper = max([__lamY(x, mu, gammas, Sx_list, influencer[i]) for x in np.linspace(0, T, num=1000)])
        Sy_list[i] = NHppp(T, lam_upper, mu, gammas, Sx_list, influencer[i])

    return Sx_list, Sy_list

File: cpp/src/test_data_network.cpp

#include "test_data_utils.h"
#include "test_data_network.h"

double TestDataNetwork::lam(double t, double mu, double* gammas, int* my_influencer, double** Sx_list, int* len_Sx_list, int N_influencer)
{
    double NY_sum = 0;

    for(int i = 0; i < N_influencer; i++)
    {
        if(my_influencer[i])
        {

            double g_sum = 0;
            for(int j = 0; j < len_Sx_list[i]; j++)
            {
                double ti = Sx_list[i][j];
                if(ti < t)
                    g_sum += g(t-Sx_list[i][j]);
            }
            NY_sum += g_sum * gammas[i];
        }

        
    }
    return mu + NY_sum;
}

void TestDataNetwork::NHppp(int T, double lam_upper, double mu, double* gammas, int* my_influencer, double** Sx_list, int* len_Sx_list, double**Sy_list, int* len_Sy_list, std::vector<double>* Sy_vec)
{
    double t = 0;
    
    while(t < T)
    {
        double u = uniform_d(mt);
        t = t - 1.0 / lam_upper * log(u);
        double u2 = uniform_d(mt);

        if(u2 < (lam(t, mu, gammas, my_influencer, Sx_list, len_Sx_list, N_influencer_heap) / lam_upper))
        {
            if(t < (double)T)
            {
                Sy_vec->emplace_back(t);
            }
        }
    }

    *Sy_list =  &(*Sy_vec)[0];
    *len_Sy_list = Sy_vec->size();

}

void TestDataNetwork::get_test_data(int T, double* lam_Sx, double mu, double* gammas, int* influencer, int N_influencer, int N_follower, double*** Sx_list, int** len_Sx_list, double*** Sy_list, int** len_Sy_list)
{
    Sx_list_heap = new double*[N_influencer];
    len_Sx_list_heap = new int[N_influencer];
    N_influencer_heap = N_influencer;
    *Sx_list = Sx_list_heap;
    *len_Sx_list = len_Sx_list_heap;

    for(int i = 0; i < N_influencer; i++){
        Hppp(T, lam_Sx[i], *Sx_list+i, *len_Sx_list+i);
    }

    Sy_list_vec_heap = new std::vector<double>[N_follower]();
    Sy_list_heap = new double*[N_follower];
    *Sy_list = Sy_list_heap;
    len_Sy_list_heap = new int[N_follower];
    *len_Sy_list = len_Sy_list_heap;

    for(int i = 0; i < N_follower; i++)
    {
        double lam_upper = 0;
        for(double x = 0; x < T; x+=(((double)T)/10000.0))
        {
            double lam_val = lam(x, mu, gammas, influencer+i*N_influencer, *Sx_list, *len_Sx_list, N_influencer);

            if(lam_val > lam_upper)
                lam_upper = lam_val;
        }
        
        NHppp(T, lam_upper, mu, gammas, influencer+i*N_influencer, *Sx_list, *len_Sx_list, *Sy_list+i, *len_Sy_list+i, Sy_list_vec_heap+i);
    }

}


TestDataNetwork::~TestDataNetwork()
{
    for(int i = 0; i < N_influencer_heap; i++)
      delete[] Sx_list_heap[i];
    delete[] Sx_list_heap;
    delete[] len_Sx_list_heap;

    
    delete[] Sy_list_vec_heap;
    delete[] len_Sy_list_heap;
    delete[] Sy_list_heap;
}


TestDataNetwork::TestDataNetwork()
{
    std::random_device rd;
    mt = std::mt19937_64(rd());
    uniform_d = std::uniform_real_distribution<double>(0,1);
}